frame/pools/src/lib.rs

```diff
@@ -0,0 +1,1135 @@
+		RewardPoolNotFound,
+		/// A sub pool does not exist.
+		SubPoolsNotFound,
+		/// An account is already delegating in another pool. An account may only belong to one
+		/// pool at a time.
+		AccountBelongsToOtherPool,
+		/// The pool has insufficient balance to bond as a nominator.
+		InsufficientBond,
+		/// The delegator is already unbonding.
+		AlreadyUnbonding,
+		/// The delegator is not unbonding and thus cannot withdraw funds.
+		NotUnbonding,
+		/// Unbonded funds cannot be withdrawn yet because the bond duration has not passed.
+		NotUnbondedYet,
+		/// The amount does not meet the minimum bond to start nominating.
+		MinimumBondNotMet,
+		/// The transaction could not be executed due to overflow risk for the pool.
+		OverflowRisk,
+		/// An error from the staking pallet.
+		StakingError,
+		// Likely only an error ever encountered in poorly built tests.
+		/// A pool with the generated account id already exists.
+		IdInUse,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Join a pre-existing pool.
+		///
+		/// Notes
+		/// * an account can only be a member of a single pool.
+		/// * this will *not* dust the delegator account, so the delegator must have at least
+		///   `existential deposit + amount` in their account.
+		#[pallet::weight(666)]
+		pub fn join(
+			origin: OriginFor<T>,
+			amount: BalanceOf<T>,
+			target: T::AccountId,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			// If a delegator already exists that means they already belong to a pool
+			ensure!(!Delegators::<T>::contains_key(&who), Error::<T>::AccountBelongsToOtherPool);
+
+			let mut bonded_pool = BondedPools::<T>::get(&target).ok_or(Error::<T>::PoolNotFound)?;
+			bonded_pool.ok_to_join_with(&target, amount)?;
+
+			// The pool should always be created in such a way its in a state to bond extra, but if
+			// the active balance is slashed below the minimum bonded or the account cannot be
+			// found, we exit early.
+			if !T::StakingInterface::can_bond_extra(&target, amount) {
+				Err(Error::<T>::StakingError)?;
+			}
+
+			// We don't actually care about writing the reward pool, we just need its
+			// total earnings at this point in time.
+			let reward_pool = RewardPools::<T>::get(&target)
+				.ok_or(Error::<T>::RewardPoolNotFound)?
+				// This is important because we want the most up-to-date total earnings.
+				.update_total_earnings_and_balance();
+
+			let old_free_balance = T::Currency::free_balance(&target);
+			// Transfer the funds to be bonded from `who` to the pools account so the pool can then
+			// go bond them.
+			T::Currency::transfer(&who, &target, amount, ExistenceRequirement::KeepAlive)?;
+
+			// Try not to bail after the above transfer
+
+			// This should now include the transferred balance.
+			let new_free_balance = T::Currency::free_balance(&target);
+			// Get the exact amount we can bond extra.
+			let exact_amount_to_bond = new_free_balance.saturating_sub(old_free_balance);
```

Comment:
Why are you assuming that the transfer function can be wrong here?

Reply:
fixed here: 3f0bbcffc67509ea11c274af71be1647c31fd8df - I wasn't sure if the currency trait made guarantees that the amount to transfer is an exact amount (and not an upper bound)